zkevm-circuits/src/mpt_circuit.rs

```diff
@@ -238,192 +226,136 @@ impl<F: Field> MPTConfig<F> {
+                // State machine
+                cb.base.set_cell_manager(state_cm.clone());
+                ifx! {f!(q_first) + f!(q_last) => {
+                    require!(a!(state_machine.is_start) => true);
+                }};
+                // Main state machine
+                matchx! {(
+                    a!(state_machine.is_start) => {
+                        state_machine.step_constraints(meta, &mut cb, StartRowType::Count as usize);
+                        cb.base.push_region(MPTRegion::Start as usize, StartRowType::Count as usize);
+                        state_machine.start_config = StartConfig::configure(meta, &mut cb, &mut ctx);
+                        ctx.memory.build_constraints(&mut cb.base, f!(q_first));
+                        cb.base.pop_region();
+                    },
+                    a!(state_machine.is_branch) => {
+                        state_machine.step_constraints(meta, &mut cb, ExtensionBranchRowType::Count as usize);
+                        cb.base.push_region(MPTRegion::Branch as usize, ExtensionBranchRowType::Count as usize);
+                        state_machine.branch_config = ExtensionBranchConfig::configure(meta, &mut cb, &mut ctx);
+                        ctx.memory.build_constraints(&mut cb.base, f!(q_first));
+                        cb.base.pop_region();
+                    },
+                    a!(state_machine.is_account) => {
+                        state_machine.step_constraints(meta, &mut cb, AccountRowType::Count as usize);
+                        cb.base.push_region(MPTRegion::Account as usize, AccountRowType::Count as usize);
+                        state_machine.account_config = AccountLeafConfig::configure(meta, &mut cb, &mut ctx);
+                        ctx.memory.build_constraints(&mut cb.base, f!(q_first));
+                        cb.base.pop_region();
+                    },
+                    a!(state_machine.is_storage) => {
+                        state_machine.step_constraints(meta, &mut cb, StorageRowType::Count as usize);
+                        cb.base.push_region(MPTRegion::Storage as usize, StorageRowType::Count as usize);
+                        state_machine.storage_config = StorageLeafConfig::configure(meta, &mut cb, &mut ctx);
+                        ctx.memory.build_constraints(&mut cb.base, f!(q_first));
+                        cb.base.pop_region();
+                    },
+                    _ => ctx.memory.build_constraints(&mut cb.base, f!(q_first)),
+                )};
+                // Only account and storage rows can have lookups, disable lookups on all other rows
+                ifx! {not!(a!(state_machine.is_account) + a!(state_machine.is_storage)) => {
+                    require!(a!(ctx.mpt_table.proof_type) => MPTProofType::Disabled.expr());
                 }}
             });
-
-            cb.base.build_constraints()
+            cb.base.build_constraints_selected(meta.query_fixed(q_enable, Rotation::cur()))
```

Comment:
I think I said it before, but I don't like this because this moves the condition to after you've read all the constraints which is very confusing, because you have no idea when the constraints are active because you only know it when you get to the end of the block of code. So you read the constraints, but don't know what `if` statement they belong to. 

It's also something that has no equivalence to normal programming language (except do while loops, but nobody uses because they are crazy!). If you program in rust/c++/typescript/python, would you ever want to move the condition of an `if` block to the end? If the code inside an `if` is too big it should be moved inside a separate function.

If users really want to (but I really don't think they should), it's also possible to achieve this without an `ifx` statement by just doing `cb.push_condition(f!(q_enable));` at the start of the block, so no special function is needed.